Sources_v3/StreamChat/APIClient/RequestEncoder.swift

```diff
@@ -81,46 +81,25 @@ struct DefaultRequestEncoder: RequestEncoder {
             }
             
             try encodeRequestBody(request: &request, endpoint: endpoint)
-            
-            // Add Stream auth headers
-            var headers = request.allHTTPHeaderFields ?? [:]
-            if let token = connectionDetailsProviderDelegate?.provideToken() {
-                headers["Stream-Auth-Type"] = "jwt"
-                headers["Authorization"] = token
-            } else {
-                headers["Stream-Auth-Type"] = "anonymous"
-            }
-            request.allHTTPHeaderFields = headers
-            
         } catch {
             completion(.failure(error))
             return
         }
-        
-        if endpoint.requiresConnectionId {
-            log.assert(
-                connectionDetailsProviderDelegate != nil,
-                "The endpoind requiers `connectionId` but `connectionDetailsProviderDelegate` is not set."
-            )
-            
-            connectionDetailsProviderDelegate?.provideConnectionId { (connectionId) in
-                guard let connectionId = connectionId else {
-                    completion(.failure(ClientError.MissingConnectionId("Failed to get `connectionId`, request can't be created.")))
-                    return
-                }
-                
-                do {
-                    request.url = try request.url?.appendingQueryItems(["connection_id": connectionId])
-                } catch {
-                    completion(.failure(error))
-                    return
+
+        addAuthorizationHeader(request: request, endpoint: endpoint) { authResult in
+            switch authResult {
+            case let .success(requestWithAuth):
+                self.addConnectionIdIfNeeded(request: requestWithAuth, endpoint: endpoint) { connectionResult in
+                    switch connectionResult {
+                    case let .success(requestWithAuthAndConnectionId):
+                        completion(.success(requestWithAuthAndConnectionId))
+                    case let .failure(error):
+                        completion(.failure(error))
+                    }
```

Comment:
```suggestion
completion(connectionResult)
```

Reply:
I wonder if there's a solution with `map`

Reply:
Hm, you mean chaining request building steps or some hashmap usage? 🤔